Document Dump/Restore protocols and a file format (#1300)
edgedb/edgedb: docs/internals/dump_format.rst

Dump File Format
================

> This description uses the same :ref:`conventions <ref_protocol_conventions>`

Reviewer: We should preface this with a warning that the dump format isn't yet stable and might change before 1.0.

Author: Added a warning

Dump File Format
================

This description uses the same :ref:`conventions <ref_protocol_conventions>`
as the protocol description.

> .. warning:: Dump file format may change until EdgeDB 1.0

Reviewer: Better wording:

```suggestion
.. warning:: The dump file format is not final and is subject to change before EdgeDB 1.0.
```

General Structure
-----------------

Dump file is structure as follows:

1. Dump file format marker ``\xFF\xD8\x00\x00\xD8EDGEDB\x00DUMP\x00``
   (17 bytes)
2. Format version number ``\x00\x00\x00\x00\x00\x00\x00\x01`` (8 bytes)
3. Header block
4. Any number of data blocks

General Dump Block
------------------

Both header and data blocks are formatted as follows:

.. code-block:: c

    struct DumpHeader {
        int8            mtype;

        // SHA1 hash sum of block data
        byte            sha1sum[20];

        // Length of message contents in bytes,
        // including self.
        int32           message_length;

>         // Data itself. Can be treated in opaque way by a client.

Reviewer: ```suggestion
        // Data itself. Should be treated in opaque way by a client.
```

        byte            data[message_length];
    }

> Usually internal structure of the block doesn't matter, all client
> need to do is:

Reviewer: ```suggestion
Upon receiving a protocol dump data message, the dump client should:
```

* Replace packet type:
    * ``@`` (0x40) → ``H`` (0x48)
    * ``=`` (0x3d) → ``D`` (0x44)
* Prepend SHA1 checksum to the block

> * Write/read the original data bytes
>   (excluding packet type byte but including the length)
> 

Reviewer: ```suggestion
* Append the entire dump protocol message disregarding the
  first byte (the message type).

```
